### ISSUE TITLE
🐛 Bug Report: Playlists fail to load after installing playlist plugin

### ISSUE THREAD

**Issue by prasoon-pxc** [backstage/backstage#14419]
### 📜 Description

I am trying to install [playlist plugin](https://github.com/backstage/backstage/tree/master/plugins/playlist-backend) for backstage, installation get successfull but when I am trying to click on playlist icon on the backstage then getting below error:
![image](https://user-images.githubusercontent.com/106815018/199648652-d514b686-d54a-43ae-8fbe-5024f87c2f46.png)
---
When I checked the backstage logs , found below error:

```
[1] 2022-11-03T04:24:57.914Z backstage info 192.168.63.1 - - [03/Nov/2022:04:24:57 +0000] "GET /api/playlist/ HTTP/1.1" 500 207 "http://192.168.63.249:3000/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36" type=incomingRequest
[1] 2022-11-03T04:24:58.147Z backstage error Cannot read properties of undefined (reading 'getIdentity') type=errorHandler stack=TypeError: Cannot read properties of undefined (reading 'getIdentity')
[1]     at evaluateRequestPermission (/home/backstage/backstage-conf-repo_my/node_modules/@backstage/plugin-playlist-backend/dist/index.cjs.js:289:33)
[1]     at /home/backstage/backstage-conf-repo_my/node_modules/@backstage/plugin-playlist-backend/dist/index.cjs.js:314:38
[1]     at handleReturn (/home/backstage/backstage-conf-repo_my/node_modules/express-promise-router/lib/express-promise-router.js:24:23)
[1]     at /home/backstage/backstage-conf-repo_my/node_modules/express-promise-router/lib/express-promise-router.js:64:7
[1]     at handleReturn (/home/backstage/backstage-conf-repo_my/node_modules/express-promise-router/lib/express-promise-router.js:24:23)
[1]     at /home/backstage/backstage-conf-repo_my/node_modules/express-promise-router/lib/express-promise-router.js:64:7
[1]     at Layer.handle [as handle_request] (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/layer.js:95:5)
[1]     at next (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/route.js:144:13)
[1]     at Route.dispatch (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/route.js:114:3)
[1]     at Layer.handle [as handle_request] (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/layer.js:95:5)
[1]     at /home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/index.js:284:15
[1]     at Function.process_params (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/index.js:346:12)
[1]     at Immediate.next (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/index.js:280:10)
[1]     at Immediate.<anonymous> (/home/backstage/backstage-conf-repo_my/node_modules/express/lib/router/index.js:646:15)
[1]     at processImmediate (node:internal/timers:468:21)
[1] 2022-11-03T04:24:58.149Z backstage info 192.168.63.1 - - [03/Nov/2022:04:24:58 +0000] "GET /api/playlist/ HTTP/1.1" 500 207 "http://192.168.63.249:3000/" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36" type=incomingRequest

```

### 👍 Expected behavior

Plugin should work as mentioned in Plugin Readme.md

### 👎 Actual Behavior with Screenshots

It should not provide any error and provide the list of playlist if it is there

### 👟 Reproduction steps

To Install the plugin on your Backstage, follow the below links
---
- First Need to install playlist-backend plugin
[Playlist Backend](https://github.com/backstage/backstage/tree/master/plugins/playlist-backend)
---
- Then install Playlist Frontend
[Playlist Frontend](https://github.com/backstage/backstage/tree/master/plugins/playlist)

### 📃 Provide the context for the Bug.

I am trying to install Playlist Plugin 

### 🖥️ Your Environment

Output of `yarn backstage-cli info`

```
yarn run v1.22.19
$ /home/backstage/backstage-conf-repo_my/node_modules/.bin/backstage-cli info
OS:   Linux 3.10.0-1160.71.1.el7.x86_64 - linux/x64
node: v16.14.0
yarn: 1.22.19
cli:  0.20.0 (installed)
backstage:  1.3.0

Dependencies:
  @backstage/app-defaults                          1.0.3
  @backstage/backend-app-api                       0.2.2
  @backstage/backend-common                        0.0.0-nightly-20220923030237, 0.14.1, 0.15.2
  @backstage/backend-plugin-api                    0.1.3
  @backstage/backend-tasks                         0.3.6
  @backstage/backend-test-utils                    0.1.29
  @backstage/catalog-client                        0.0.0-nightly-20220923030237, 0.5.5, 1.1.1
  @backstage/catalog-model                         0.0.0-nightly-20220923030237, 0.9.10, 1.1.2
  @backstage/cli-common                            0.1.10
  @backstage/cli                                   0.17.2, 0.20.0
  @backstage/config-loader                         0.0.0-nightly-20220923030237, 1.1.5
  @backstage/config                                0.0.0-nightly-20220923030237, 0.1.15, 1.0.3
  @backstage/core-app-api                          1.0.3
  @backstage/core-components                       0.0.0-nightly-2022122206, 0.10.0, 0.11.2, 0.9.5
  @backstage/core-plugin-api                       0.0.0-nightly-20220923030237, 0.6.1, 1.0.7
  @backstage/errors                                1.1.0, 0.0.0-nightly-20220923030237, 0.2.2, 1.1.2
  @backstage/integration-react                     1.1.2
  @backstage/integration                           0.0.0-nightly-20220923030237, 0.7.5, 1.3.2
  @backstage/plugin-api-docs                       0.8.6
  @backstage/plugin-app-backend                    0.3.33
  @backstage/plugin-auth-backend                   0.14.1
  @backstage/plugin-auth-node                      0.2.6
  @backstage/plugin-catalog-backend-module-github  0.1.5
  @backstage/plugin-catalog-backend                1.3.0
  @backstage/plugin-catalog-common                 1.0.7
  @backstage/plugin-catalog-graph                  0.2.18
  @backstage/plugin-catalog-import                 0.8.9
  @backstage/plugin-catalog-node                   1.0.0
  @backstage/plugin-catalog-react                  0.0.0-nightly-2022122206, 1.2.0
  @backstage/plugin-catalog                        1.3.0
  @backstage/plugin-github-actions                 0.5.6
  @backstage/plugin-home                           0.4.23
  @backstage/plugin-org                            0.5.6
  @backstage/plugin-permission-common              0.4.0, 0.5.3, 0.6.4, 0.7.0
  @backstage/plugin-permission-node                0.6.3, 0.7.0
  @backstage/plugin-permission-react               0.3.4, 0.4.6
  @backstage/plugin-playlist-backend               0.2.0
  @backstage/plugin-playlist-common                0.1.1
  @backstage/plugin-playlist                       0.1.1
  @backstage/plugin-proxy-backend                  0.2.27
  @backstage/plugin-scaffolder-backend             1.3.0
  @backstage/plugin-scaffolder-common              1.1.2
  @backstage/plugin-scaffolder                     1.3.0
  @backstage/plugin-search-backend-module-pg       0.3.4
  @backstage/plugin-search-backend-node            0.6.2
  @backstage/plugin-search-backend                 0.5.3
  @backstage/plugin-search-common                  0.3.5, 1.1.0
  @backstage/plugin-search-react                   0.2.1, 1.2.0
  @backstage/plugin-search                         0.9.0
  @backstage/plugin-shortcuts                      0.3.2
  @backstage/plugin-stack-overflow                 0.1.3
  @backstage/plugin-tech-radar                     0.5.13
  @backstage/plugin-techdocs-backend               1.1.2
  @backstage/plugin-techdocs-module-addons-contrib 1.0.1
  @backstage/plugin-techdocs-node                  1.1.2
  @backstage/plugin-techdocs-react                 1.0.1
  @backstage/plugin-techdocs                       1.2.0
  @backstage/plugin-todo-backend                   0.0.0-nightly-20220923030237, 0.1.33
  @backstage/plugin-todo                           0.0.0-nightly-20220923030237
  @backstage/plugin-user-settings                  0.4.5
  @backstage/release-manifests                     0.0.4, 0.0.6
  @backstage/test-utils                            1.1.1
  @backstage/theme                                 0.2.16
  @backstage/types                                 0.1.3, 1.0.0
  @backstage/version-bridge                        0.1.2, 1.0.1
Done in 0.96s.
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by kuangp** [comment]
Hi @prasoon-pxc, can you share what your backend playlist config looks like eg. your `packages/backend/src/plugins/playlist.ts` file?

**Comment by prasoon-pxc** [comment]
Hi @kuangp , Please find the files below:

`playlist.ts`
```
import { createRouter } from '@backstage/plugin-playlist-backend';
import { Router } from 'express';
import { PluginEnvironment } from '../types';

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  return await createRouter({
    database: env.database,
    discovery: env.discovery,
    logger: env.logger,
    permissions: env.permissions,
  });
}
```
---
`permissions.ts`

```
import { createRouter } from '@backstage/plugin-permission-backend';
import {
  AuthorizeResult,
  PolicyDecision,
} from '@backstage/plugin-permission-common';
import { PermissionPolicy } from '@backstage/plugin-permission-node';
import { Router } from 'express';
import { PluginEnvironment } from '../types';

class TestPermissionPolicy implements PermissionPolicy {
  async handle(): Promise<PolicyDecision> {
    return { result: AuthorizeResult.ALLOW };
  }
}

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  return await createRouter({
    config: env.config,
    logger: env.logger,
    discovery: env.discovery,
    policy: new TestPermissionPolicy(),
    identity: env.identity,
  });
}
```
---
 Part of `index.ts`
```
import playlist from './plugins/playlist';
const playlistEnv = useHotMemoize(module, () => createEnv('playlist'));

  const apiRouter = Router();
  apiRouter.use('/playlist', await playlist(playlistEnv));
```

**Comment by kuangp** [comment]
It looks like you forgot to pass in the `IdentityApi` to the plugin config as noted in the [setup instructions](https://github.com/backstage/backstage/tree/master/plugins/playlist-backend#adding-the-plugin-to-your-packagesbackend):

```diff
export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  return await createRouter({
    database: env.database,
    discovery: env.discovery,
+  identity: env.identity,
    logger: env.logger,
    permissions: env.permissions,
  });
}
```

Can you try to make the above change and see if this resolves your issue?

**Comment by Rugvip** [comment]
Gonna close since @kuangp's suggestion should fix the issue. Please let us know if it doesn't

**Comment by prasoon-pxc** [comment]
@kuangp , @Rugvip Now , I have added the `IdentityApi` as you have mentioned but getting the same error 

`playlist.ts`
---
```
import { createRouter } from '@backstage/plugin-playlist-backend';
import { Router } from 'express';
import { PluginEnvironment } from '../types';

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  return await createRouter({
    database: env.database,
    discovery: env.discovery,
    logger: env.logger,
    identity: env.identity,
    permissions: env.permissions,
  });
}
```

**Comment by kuangp** [comment]
@prasoon-pxc it looks like you're on version `1.3.0` of Backstage which is pretty outdated now. Can you try [updating](https://backstage.io/docs/getting-started/keeping-backstage-updated) to the latest version and see if that resolves your issue?

**Comment by prasoon-pxc** [comment]
@kuangp --> yesterday, I had also tried with updating the backstage version, but with no luck.
---
`yarn backstage-cli info`

---
```
backstage@backstage backstage-conf-repo_my]$ yarn backstage-cli info
yarn run v1.22.19
$ /home/backstage/backstage-conf-repo_my/node_modules/.bin/backstage-cli info
OS:   Linux 3.10.0-1160.71.1.el7.x86_64 - linux/x64
node: v16.14.0
yarn: 1.22.19
cli:  0.20.0 (installed)
backstage:  1.7.1

Dependencies:
  @backstage/app-defaults                          1.0.7
  @backstage/backend-app-api                       0.2.2
  @backstage/backend-common                        0.15.2
  @backstage/backend-plugin-api                    0.1.3
  @backstage/backend-tasks                         0.3.6
  @backstage/backend-test-utils                    0.1.29
  @backstage/catalog-client                        1.1.1
  @backstage/catalog-model                         1.1.2
  @backstage/cli-common                            0.1.10
  @backstage/cli                                   0.20.0
  @backstage/config-loader                         1.1.5
  @backstage/config                                1.0.3
  @backstage/core-app-api                          1.1.1
  @backstage/core-components                       0.10.0, 0.11.2
  @backstage/core-plugin-api                       1.0.7
  @backstage/errors                                1.1.2
  @backstage/integration-react                     1.1.5
  @backstage/integration                           1.3.2
  @backstage/plugin-api-docs                       0.8.10
  @backstage/plugin-app-backend                    0.3.37
  @backstage/plugin-auth-backend                   0.17.0
  @backstage/plugin-auth-node                      0.2.6
  @backstage/plugin-bazaar-backend                 0.2.0
  @backstage/plugin-bazaar                         0.1.25
  @backstage/plugin-catalog-backend-module-github  0.1.8
  @backstage/plugin-catalog-backend                1.5.0
  @backstage/plugin-catalog-common                 1.0.7
  @backstage/plugin-catalog-graph                  0.2.22
  @backstage/plugin-catalog-import                 0.9.0
  @backstage/plugin-catalog-node                   1.2.0
  @backstage/plugin-catalog-react                  1.2.0
  @backstage/plugin-catalog                        1.6.0
  @backstage/plugin-gcalendar                      0.3.6
  @backstage/plugin-github-actions                 0.5.10
  @backstage/plugin-home                           0.4.26
  @backstage/plugin-org                            0.5.10
  @backstage/plugin-permission-common              0.7.0
  @backstage/plugin-permission-node                0.7.0
  @backstage/plugin-permission-react               0.4.6
  @backstage/plugin-playlist-backend               0.2.0
  @backstage/plugin-playlist-common                0.1.1
  @backstage/plugin-playlist                       0.1.1
  @backstage/plugin-proxy-backend                  0.2.31
  @backstage/plugin-scaffolder-backend             1.7.0
  @backstage/plugin-scaffolder-common              1.2.1
  @backstage/plugin-scaffolder                     1.7.0
  @backstage/plugin-search-backend-module-pg       0.4.1
  @backstage/plugin-search-backend-node            1.0.3
  @backstage/plugin-search-backend                 1.1.0
  @backstage/plugin-search-common                  1.1.0
  @backstage/plugin-search-react                   1.2.0
  @backstage/plugin-search                         1.0.3
  @backstage/plugin-shortcuts                      0.3.2
  @backstage/plugin-stack-overflow                 0.1.6
  @backstage/plugin-tech-radar                     0.5.17
  @backstage/plugin-techdocs-backend               1.4.0
  @backstage/plugin-techdocs-module-addons-contrib 1.0.5
  @backstage/plugin-techdocs-node                  1.4.1
  @backstage/plugin-techdocs-react                 1.0.5
  @backstage/plugin-techdocs                       1.3.3
  @backstage/plugin-todo-backend                   0.1.34
  @backstage/plugin-todo                           0.2.12
  @backstage/plugin-user-settings                  0.5.0
  @backstage/release-manifests                     0.0.6
  @backstage/test-utils                            1.2.1
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
Done in 0.80s.
```

**Comment by kuangp** [comment]
do you mind sharing what the rest of your `packages/backend/src/index.ts` looks like? particularly the `makeCreateEnv` function. Does it match up with the [default template](https://github.com/backstage/backstage/blob/master/packages/create-app/templates/default-app/packages/backend/src/index.ts#L35-L72) where the `IdentityApi` should be defined?

**Comment by prasoon-pxc** [comment]
@kuangp --> I checked the my `index.ts` and `types.ts` and did not find it updated with `default template`  ,
   After updating below files, issue got resolved. 

---
- Need to add below line in `index.ts`

 ```
import { DefaultIdentityClient } from '@backstage/plugin-auth-node';

function makeCreateEnv(config: Config) {
  const identity = DefaultIdentityClient.create({
    discovery,
  });
...
return (plugin: string): PluginEnvironment => {

   return {
     ...
      identity,

```
- And in `packages/backend/src/types.ts`
```
import { IdentityApi } from '@backstage/plugin-auth-node';

export type PluginEnvironment = {
 ...
  identity: IdentityApi;
};

```
---
Thanks for your time

**Comment by regicsolutions** [comment]
@kuangp i followed all steps here and am running into the following error when i login as guest, i have permissions completely disabled, any thoughts on this one?

`{"error":{"name":"NotAllowedError","message":"Unauthorized","stack":"NotAllowedError: Unauthorized\n    at evaluateRequestPermission (appl\\node_modules\\@backstage\\plugin-playlist-backend\\dist\\index.cjs.js:291:13)\n    
at async app\\node_modules\\@backstage\\plugin-playlist-backend\\dist\\index.cjs.js:314:32"},"request":{"method":"GET","url":"/"},"response":{"statusCode":403}}`

![image](https://user-images.githubusercontent.com/6157895/201237909-d56b2fd3-0d8f-4bb0-9da8-251fda007c57.png)

**Comment by kuangp** [comment]
hmm unfortunately I don't think guest users are supported by this plugin since even though it doesn't necessarily require permissions to be enabled, the playlist backend does expect requests to contain an [id token](https://github.com/backstage/backstage/blob/af5d2741da4900310b439d8a0158f3a2e7e5e45f/plugins/playlist-backend/src/service/router.ts#L84-L88) so it can extract the user identity (to determine playlist followings) which the [default guest provider](https://github.com/backstage/backstage/blob/afa90a3e5148ad955d8fab545177d18988e05ab5/packages/core-components/src/layout/SignInPage/GuestUserIdentity.ts#L28-L30) does not produce

**Comment by prasoon-pxc** [comment]
@regicsolutions --> You can look through this link to enable Authentication provider , by this you will get id token on sign-in and hopefully your playlist plugin will work then.
[Authentication-Providers](https://backstage.io/docs/auth/#configuring-authentication-providers)

**Comment by jamieklassen** [comment]
I think @kuangp's observation would be resolved by #12394
